README.rst
==================
 nti.recipes.zodb
==================

Opinionated recipes for creating RelStorage and ZEO configurations, especially
tailored for multi-databases.

.. image:: https://travis-ci.org/NextThought/nti.recipes.zodb.svg?branch=master
    :target: https://travis-ci.org/NextThought/nti.recipes.zodb

.. image:: https://coveralls.io/repos/github/NextThought/nti.recipes.zodb/badge.svg?branch=master
   :target: https://coveralls.io/github/NextThought/nti.recipes.zodb?branch=master

Limitations
===========

A single buildout can use *either* one RelStorage recipe *or* one ZEO
recipe. It can never have both, or more than one of each. This is
because both recipes write to the same configuration files

Dependencies
============

The recipes defined here use `collective.recipe.template`_ to output
configuration files, and `z3c.recipe.mkdir`_ to create implicitly
defined directories. `zc.zodbrecipes`_ is used to create the ZEO
server. You shouldn't need to install these manually as buildout will
take care of making them available when needed.

.. _collective.recipe.template: https://pypi.org/project/collective.recipe.template/
.. _z3c.recipe.mkdir: https://pypi.org/project/z3c.recipe.mkdir/
.. _zc.zodbrecipes: https://pypi.org/project/zc.zodbrecipes/

Directories and Files
=====================

The recipes defined here use the directory structure and variables
defined by `zc.recipe.deployment`_. There should be a buildout part
called ``deployment`` that uses this recipe. Alternatively (and
especially useful when composing a buildout from multiple
configurations), you can define a ``deployment`` part that lists these
directories manually. You'll also need to include the username that
should own created files and directories:

.. code:: ini

   [deployment]
   bin-directory = ${buildout:bin-directory}
   cache-directory = ${:run-directory}/caches
   crontab-directory = ${:root-directory}/etc/cron.d
   data-directory = ${:root-directory}/data
   etc-directory = ${:root-directory}/etc
   log-directory = ${:root-directory}/var/log
   logrotate-directory = ${:root-directory}/etc/logrotate.d
   rc-directory = ${:root-directory}/bin/rc
   root-directory = ${buildout:root-directory}
   run-directory = ${:root-directory}/var

The deployment recipe takes care of creating the needed directories,
but here we'll just do so manually. We'll define a common
configuration snippet that we'll include in future examples::

    >>> write(sample_buildout, 'deployment.cfg',
    ... """
    ... [deployment]
    ... root-directory = ${buildout:directory}
    ... data-directory = ${:root-directory}/data
    ... etc-directory = ${:root-directory}/etc
    ... log-directory = ${:root-directory}/var/log
    ... run-directory = ${:root-directory}/var
    ... rc-directory = ${:root-directory}/bin/rc
    ... cache-directory = ${:run-directory}/caches
    ... logrotate-directory = ${:root-directory}/etc/logrotate.d
    ... crontab-directory = ${:root-directory}/etc/cron.d
    ... user = user
    ...
    ... [directories]
    ... recipe = z3c.recipe.mkdir
    ... create-intermediate = true
    ... mode = 0700
    ... paths =
    ...    ${deployment:etc-directory}
    ...    ${deployment:run-directory}
    ...    ${deployment:cache-directory}
    ...    ${deployment:data-directory}
    ...    ${deployment:log-directory}
    ...    ${deployment:rc-directory}
    ...    ${deployment:logrotate-directory}
    ...    ${deployment:crontab-directory}
    ... """)

Both recipes create two files in the ``etc-directory``.

zodb_conf.xml
    This file is meant to be read with
    ``ZODB.config.databaseFromFile`` or ``databaseFromURL``. If you
    specify more than one storage, they will be listed in the order
    provided, creating a multi-database, with the first listed storage
    as the "root" database.

zeo_uris.ini
    This file provides the same database configuration as
    ``zodb_conf.xml`` (indeed, it references that file), but in a form
    of a single URL string that can be read using zodburi_. This can
    be convient for passing in the form of a string.

.. _zc.recipe.deployment: https://pypi.org/project/zc.recipe.deployment/_
.. _zodburi: https://pypi.org/project/zodburi/


Recipe Options
==============

Both recipes defined here accept some common options.

storages
    Required. A whitespace delimited list of storage names. Each of these will be
    added to the generated configuration files for a client to use (and
    for ZEO, for the server to serve).

    This can only be defined directly in the recipe part.
compress
   If "decompress" (the default) each storage will be wrapped in a
   `zc.zlibstorage`_ that only compress existing records. If set to
   "compress" new records will also be compressed.

   Set to "none" to disable the wrapper entirely.

   This can be set in the recipe part. If it's not defined there, a
   value defined in the ``environment`` part will be used before
   falling back to the default.

.. _zc.zlibstorage: https://pypi.org/project/zc.zlibstorage/

Storage and Database Options
============================

Some options are available to configure the ZODB database. These are
used by both recipes and may be defined at a per-database level (see
each recipe for an explanation of how). The defaults are built-in, but
setting a value in the recipe part will provide a new default for all
storages. Additionally, for backwards compatibility and composing
buildout configurations, if there is a part named ``<part>_opts``,
where ``<part>`` is the name of the recipe part, options defined there
will override options defined in the recipe part, but will bee
overridden by options defined for an individual storage.

These configuration options have to do with the ZODB connection pool
and its caching.

.. code:: ini

   [zeo]
   recipe = nti.recipes.zodb:zeo
   storages = users
   cache_size = 50

.. code:: ini

   [zeo]
   recipe = nti.recipes.zodb:zeo
   storages = users

   [zeo_opts]
   cache_size = 50


cache_size
   Controls the ZODB per-connection object cache. Setting this to a large-enough
   value to contain your application's working set can be very important, especially
   in read-heavy workloads. Setting it too large can waste memory.
pool_size
    Controls the number of ZODB connections kept in the ZODB pool. It
    is very important to set this large enough to accomodate the
    number of concurrent activities (requests) your application will
    handle. Each connection in the pool holds resources like its cache
    and in the case of RelStorage RDBMS sockets and possibly memcache
    sockets. Setting it too large can waste memory and file-descriptors.

    Normally, opening a DB and closing the connection will create a
    connection (if needed), then return it to the pool (if the pool is
    not full). However, in the case of multi-databases, when an object
    from a secondary database needs to be loaded, the active
    connection will request a connection to that database, and when
    the active connection is closed, that secondary connection is also
    closed *but not returned to the pool*. Instead, the active
    (primary) connection keeps a reference to it that it will use in
    the future. This has the effect of driving all secondary pools
    based on the efficiency of the primary pool. Thus, the pool-size
    for everything except the primary database is essentially
    meaningless (if the application always begins by opening that
    primary database), but that pool size controls everything.

    Calling DB.connectionDebugInfo() can show improperly sized pools:
    connections in the pool have 'opened' of None, while those in use
    have a timestamp and the length of time it's been open.
pool_timeout
    A time interval value (which accepts either a bare number of
    integral seconds or an integer suffixed with one of the characters
    's', 'm', 'h', 'd' for seconds, minutes, hours or days,
    respectively) that specifies how long idle connections are allowed
    to remain in the pool before being closed. Effectively, there is
    no default meaning connections never time out.


RelStorage
==========

The ``relstorage`` recipe creates configurations to connect to a
MySQL, PostgreSQL, or SQLite3 RelStorage database *in history-free
mode*. (Oracle is not supported.)

There are a number of different ways to configure storages. If you
have multiple storages residing on a common database server, and you
also have other databases on that server (SQLAlchemy, etc), you might
be interested in using a shared ``environment`` section to contain
the server location and account credentials.

.. note:: Do not store plain passwords in buildout configuration
          files. Use something like `nti.recipes.passwords
          <https://pypi.org/project/nti.recipes.passwords/>`_ to store
          them encrypted instead.

By default, the name of the database is the same as the storage name.
The storage name's case is preserved (this matters for cross-database
references.)

    >>> write(sample_buildout, 'buildout.cfg',
    ... """
    ... [buildout]
    ... extends = deployment.cfg
    ... parts = directories relstorage
    ...
    ... [environment]
    ... sql_user = the_user
    ... sql_host = the_server
    ... sql_passwd = the_passwd
    ...
    ... [relstorage]
    ... recipe = nti.recipes.zodb:relstorage
    ... storages = Users Sessions
    ... compress = none
    ... """)

    >>> print_(system(buildout), end='')
    Installing...
    Installing relstorage.
    >>> ls(sample_buildout, 'etc')
    d  cron.d
    d  logrotate.d
    -  zeo_uris.ini
    -  zodb_conf.xml
    >>> cat(sample_buildout, 'etc', 'zodb_conf.xml')
    %import relstorage
    <zodb Users>
      cache-size 100000
      database-name Users
      pool-size 60
      <relstorage Users>
        <mysql>
          # This comment preserves whitespace
              db Users
              host the_server
              passwd the_passwd
              user the_user
        </mysql>
      blob-dir /sample-buildout/data/Users.blobs
      cache-local-dir /sample-buildout/var/caches/data_cache/Users.cache
      cache-local-mb 300
      cache-prefix Users
      commit-lock-timeout 60
      keep-history false
      name Users
      pack-gc false
      shared-blob-dir false
    </relstorage>
    </zodb>
    <zodb Sessions>
      cache-size 100000
      database-name Sessions
      pool-size 60
      <relstorage Sessions>
        <mysql>
          # This comment preserves whitespace
              db Sessions
              host the_server
              passwd the_passwd
              user the_user
        </mysql>
      blob-dir /sample-buildout/data/Sessions.blobs
      cache-local-dir /sample-buildout/var/caches/data_cache/Sessions.cache
      cache-local-mb 300
      cache-prefix Sessions
      commit-lock-timeout 60
      keep-history false
      name Sessions
      pack-gc false
      shared-blob-dir false
    </relstorage>
    </zodb>
    >>> cat(sample_buildout, 'etc', 'zeo_uris.ini')
    [ZODB]
    uris = /sample-buildout/etc/zodb_conf.xml#users /sample-buildout/etc/zodb_conf.xml#sessions

Much can be configured at both the recipe level and in a section named
for the storage (prefixed with the name of the recipe, unlike in the
zeo recipe, and suffixed with '_storage_opts'), but a few settings can
only be configured on the recipe or environment part.

enable-persistent-cache
    Defaults to true.
cache-servers
   Deprecated. A list of memcache servers to use. Can be configured at the recipe
   or environment level.
blob-cache-size
   Defaults to no size cap.

    >>> write(sample_buildout, 'buildout.cfg',
    ... """
    ... [buildout]
    ... extends = deployment.cfg
    ... parts = directories relstorage
    ...
    ... [environment]
    ... sql_user = the_user
    ... sql_host = the_server
    ... sql_passwd = the_passwd
    ...
    ... [relstorage]
    ... recipe = nti.recipes.zodb:relstorage
    ... storages = users sessions
    ... compress = none
    ... pack_gc = true
    ... commit_lock_timeout = 10
    ... pool_timeout = 42s
    ...
    ... [relstorage_users_storage_opts]
    ... cache_size = 42
    ... cache_local_mb = 2
    ... sql_user = custom_user
    ... """)

    >>> print_(system(buildout), end='')
    Uninstalling relstorage...
    Installing relstorage.
    >>> ls(sample_buildout, 'etc')
    d  cron.d
    d  logrotate.d
    -  zeo_uris.ini
    -  zodb_conf.xml
    >>> cat(sample_buildout, 'etc', 'zodb_conf.xml')
    %import relstorage
    <zodb users>
      cache-size 42
      database-name users
      pool-size 60
      pool-timeout 42s
      <relstorage users>
        <mysql>
          # This comment preserves whitespace
              db users
              host the_server
              passwd the_passwd
              user custom_user
        </mysql>
      blob-dir /sample-buildout/data/users.blobs
      cache-local-dir /sample-buildout/var/caches/data_cache/users.cache
      cache-local-mb 2
      cache-prefix users
      commit-lock-timeout 10
      keep-history false
      name users
      pack-gc true
      shared-blob-dir false
    </relstorage>
    </zodb>
    <zodb sessions>
      cache-size 100000
      database-name sessions
      pool-size 60
      pool-timeout 42s
      <relstorage sessions>
        <mysql>
          # This comment preserves whitespace
              db sessions
              host the_server
              passwd the_passwd
              user the_user
        </mysql>
      blob-dir /sample-buildout/data/sessions.blobs
      cache-local-dir /sample-buildout/var/caches/data_cache/sessions.cache
      cache-local-mb 300
      cache-prefix sessions
      commit-lock-timeout 10
      keep-history false
      name sessions
      pack-gc true
      shared-blob-dir false
    </relstorage>
    </zodb>

Configuring The Adapter
-----------------------

By default, a MySQL adapter is assumed. Use the ``sql_adapter``
setting (at the recipe or storage level) to change this.

The ``sql_adapter_extra_args`` may be used to add additional
configuration to the ``<adapter>`` section. This is frequently used to
select a driver.

If you change it to ``postgresql`` a DSN will be constructed based on
the ``sql_*`` settings. You can set ``sql_adapter_args`` to completely
specify the contents of the ``<adapter>`` section (this disables
``sql_adapter_extra_args``).

    >>> write(sample_buildout, 'buildout.cfg',
    ... """
    ... [buildout]
    ... extends = deployment.cfg
    ... parts = directories relstorage
    ...
    ... [environment]
    ... sql_user = the_user
    ... sql_host = the_server
    ... sql_passwd = the_passwd
    ...
    ... [relstorage]
    ... recipe = nti.recipes.zodb:relstorage
    ... storages = users sessions
    ... compress = none
    ... sql_adapter = postgresql
    ... sql_adapter_extra_args =
    ...     driver gevent psycopg2
    ...
    ... [relstorage_users_storage_opts]
    ... sql_adapter = sqlite3
    ... sql_adapter_extra_args =
    ...     driver gevent sqlite3
    ... """)

    >>> print_(system(buildout), end='')
    Uninstalling relstorage...
    Installing relstorage.
    >>> ls(sample_buildout, 'etc')
    d  cron.d
    d  logrotate.d
    -  zeo_uris.ini
    -  zodb_conf.xml
    >>> cat(sample_buildout, 'etc', 'zodb_conf.xml')
    %import relstorage
    <zodb users>
      cache-size 100000
      database-name users
      pool-size 60
      <relstorage users>
        <sqlite3>
          # This comment preserves whitespace
              data-dir /sample-buildout/data/relstorage_users_storage
              driver gevent sqlite3
        </sqlite3>
      blob-dir /sample-buildout/data/users.blobs
      cache-local-dir /sample-buildout/var/caches/data_cache/users.cache
      cache-local-mb 300
      cache-prefix users
      commit-lock-timeout 60
      keep-history false
      name users
      pack-gc false
      shared-blob-dir true
    </relstorage>
    </zodb>
    <zodb sessions>
      cache-size 100000
      database-name sessions
      pool-size 60
      <relstorage sessions>
        <postgresql>
          # This comment preserves whitespace
              driver gevent psycopg2
              dsn dbname='sessions' user='the_user' password='the_passwd' host='the_server'
        </postgresql>
      blob-dir /sample-buildout/data/sessions.blobs
      cache-local-dir /sample-buildout/var/caches/data_cache/sessions.cache
      cache-local-mb 300
      cache-prefix sessions
      commit-lock-timeout 60
      keep-history false
      name sessions
      pack-gc false
      shared-blob-dir false
    </relstorage>
    </zodb>

Other Files
-----------

If the recipe was ``write-zodbconvert`` set to ``true``, then a set of
configuration files for converting to and from RelStorage and
FileStorage using ``zodbconvert`` will be generated.

    >>> write(sample_buildout, 'buildout.cfg',
    ... """
    ... [buildout]
    ... extends = deployment.cfg
    ... parts = directories relstorage
    ...
    ... [environment]
    ... sql_user = the_user
    ... sql_host = the_server
    ... sql_passwd = the_passwd
    ...
    ... [relstorage]
    ... recipe = nti.recipes.zodb:relstorage
    ... storages = users
    ... compress = none
    ... write-zodbconvert = true
    ... """)

    >>> print_(system(buildout), end='')
    Uninstalling relstorage...
    Installing relstorage.
    >>> ls(sample_buildout, 'etc')
    d  cron.d
    d  logrotate.d
    d  relstorage
    -  zeo_uris.ini
    -  zodb_conf.xml
    >>> ls(sample_buildout, 'etc', 'relstorage')
    -  users_from_relstorage_conf.xml
    -  users_to_relstorage_conf.xml

ZEO
===

The ``zeo`` recipe can be used to create configurations for a ZEO
client and ZEO server. It is only intended for personal or test
environments.

.. rubric:: Options

Just like the ``relstorage`` recipe, it requires one or more storages.
Options can be set in the ``zeo`` part to apply to all storages, in a
part named for the storage to configure the server, or in a part named
for the client to configure the client. Note that the client also
inherits configuration options from the server.

pack-gc
   Defaults to false. This can only be set on the recipe part.


    >>> write(sample_buildout, 'buildout.cfg',
    ... """

CHANGES.rst
=========
 Changes
=========

1.1.1 (unreleased)
==================

- Nothing changed yet.


1.1.0 (2020-10-06)
==================

- Add support for Python 3.9.

- Add support for configuring the ZODB ``pool-timeout`` value. See
  `issue 19
  <https://github.com/NextThought/nti.recipes.zodb/issues/19>`_.


1.0.0a3 (2019-11-18)
====================

- ZEO: Fix a mismatch in case and storage names for the client ZCML in
  zodb_conf.xml. See `issue 17 <https://github.com/NextThought/nti.recipes.zodb/issues/17>`_.


1.0.0a2 (2019-11-15)
====================

- RelStorage: The value of ``sql_adapter_extra_args`` is validated to
  be syntactically correct.

- RelStorage: Support providing a PostgreSQL DSN

- All storages: Change the default to use zlibstorage only to
  decompress existing records, not to compress new records. Set the
  ``compress`` option (in this recipe or the ``environment`` recipe)
  to ``compress`` to turn compression on. Set it to ``decompress`` to
  explicitly request only decompression, and set it to ``none`` to
  explicitly disable all compression and decompression. In the future,
  expect the default to change to ``none``. See `issue 9 <https://github.com/NextThought/nti.recipes.zodb/issues/9>`_.

- ZEO: Instead of using ``${buildout:directory}/var/`` and
  ``${buildout:directory}/var/log`` directly, refer
  to ``${deployment:run-directory}`` and ``${deployment:log-directory}``.

- All storages: Make previously hard-coded values configurable. This
  includes ``pool-size`` (``pool_size``), ``commit-lock-timeout``
  (``commit_lock_timeout_``) and ``cache-size`` (``cache_size``).
  These values can be set in the recipe, in the ``_opts`` section, or
  in the ``_opts`` section for a particular storage.

- RelStorage PostgreSQL: Automatically use a correct DSN based on the
  configured host, port, dbname and password if no ``dsn`` setting is
  specified. Previously it was complex to specify a DSN using
  configurable values.

- RelStorage: Make ``shared_blob_dir``, ``cache_local_mb`` and
  ``pack_gc`` be configurable at the storage level instead of just the
  recipe part level.

1.0.0a1 (2019-11-14)
====================

- Fix relstorage recipe on Python 3. See `issue 8
  <https://github.com/NextThought/nti.recipes.zodb/issues/8>`_.

- For RelStorage, if the ``sql_adapter`` is set to ``sqlite3``, then
  derive a path to the data directory automatically. This can be set
  at the main part level, the part_opts level, or the
  part_storage_opts level. Also automatically set ``shared-blob-dir``
  to true.

- For RelStorage, avoid writing out the deprecated
  ``cache-local-dir-count`` option.


src/nti/recipes/zodb/relstorage.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
A meta-recipe to create multiple
relstorage connections in a Dataserver buildout.

"""

from __future__ import print_function
from __future__ import absolute_import
from __future__ import division

import io

import ZConfig.schemaless

from ._model import Part
from ._model import ZConfigSection
from ._model import ZConfigSnippet
from ._model import Ref as SubstVar
from ._model import RelativeRef as LocalSubstVar
from ._model import hyphenated
from ._model import Default

from . import MultiStorageRecipe
from . import filestorage
from . import zodb
from . import ZodbClientPart

logger = __import__('logging').getLogger(__name__)
NativeStringIO = io.BytesIO if bytes is str else io.StringIO

def _option_true(value):
    return value and value.lower() in ('1', 'yes', 'on', 'true')

class relstorage(ZConfigSection):
    blob_cache_size = LocalSubstVar('blob-cache-size').hyphenate()
    blob_dir = LocalSubstVar("blob_dir").hyphenate()
    cache_local_dir = LocalSubstVar('cache-local-dir').hyphenate()
    cache_local_mb = LocalSubstVar('cache-local-mb').hyphenate()
    cache_prefix = LocalSubstVar('name').hyphenate()
    commit_lock_timeout = LocalSubstVar('commit_lock_timeout').hyphenate()
    keep_history = hyphenated(False)
    name = LocalSubstVar('relstorage-name-prefix') + LocalSubstVar('name')
    pack_gc = LocalSubstVar('pack-gc').hyphenate()
    shared_blob_dir = LocalSubstVar('shared-blob-dir').hyphenate()

    def __init__(self, memcache_config):
        ZConfigSection.__init__(
            self, 'relstorage', LocalSubstVar('name'),
            # One section, <$adapter>
            ZConfigSection(
                LocalSubstVar('sql_adapter'), None,
                APPEND=LocalSubstVar('sql_adapter_args')
            ),
            APPEND=memcache_config,
        )

class BaseStoragePart(ZodbClientPart):
    blob_cache_size = Default(None).hyphenate()
    blob_dir = LocalSubstVar('data_dir') / LocalSubstVar('name') + '.blobs'
    blob_dump_dir = (
        LocalSubstVar('data_dir')
        / 'relstorage_dump'
        / LocalSubstVar('dump_name')
        / 'blobs'
    )
    cache_local_dir = hyphenated(None)
    cache_local_mb = Default(300).hyphenate()

    commit_lock_timeout = Default(60)
    data_dir = SubstVar('deployment', 'data-directory')
    dump_dir = LocalSubstVar('data_dir') / 'relstorage_dump' / LocalSubstVar('dump_name')
    dump_name = LocalSubstVar('name')
    filestorage_name = 'NONE'
    name = 'BASE'
    pack_gc = Default(False).hyphenate()
    relstorage_name_prefix = hyphenated(None)
    # Prior to RelStorage 3, by default, relstorage assumes a shared blob
    # directory. However, our most common use case here
    # is not to share. While using either wrong setting
    # in an environment is dangerous and can lead to data loss,
    # it's slightly worse to assume shared when its not
    shared_blob_dir = Default(False).hyphenate()

    sql_db = LocalSubstVar('name')
    sql_adapter_args = ZConfigSnippet(
        db=LocalSubstVar('sql_db'),
        user=LocalSubstVar('sql_user'),
        passwd=LocalSubstVar('sql_passwd'),
        host=LocalSubstVar('sql_host'),
        APPEND=LocalSubstVar('sql_adapter_extra_args')
    )
    sql_adapter_extra_args = None


def _ZConfig_write_to(config, writer, part):
    writer.begin_line("# This comment preserves whitespace")
    indent = writer.current_indent * 2 + '  '
    for line in config.__str__(indent).splitlines():
        writer.begin_line(line)

ZConfig.schemaless.Section.write_to = _ZConfig_write_to

class Databases(MultiStorageRecipe):

    def __init__(self, buildout, name, options):
        MultiStorageRecipe.__init__(self, buildout, name, options)
        # Get the 'environment' block from buildout if it exists. This is for
        # combatibility with existing buildouts.
        environment = buildout.get('environment', {})
        relstorage_name_prefix = options.get('relstorage-name-prefix', '')

        # The initial use case has the same SQL database, SQL user,
        # cache servers, etc, for all connections. Using _opts sections
        # either for the this element or for an individual storage in this element
        # can override it.
        sql_user = options.get('sql_user') or environment.get('sql_user')
        sql_passwd = options.get('sql_passwd') or environment.get('sql_passwd')
        sql_host = options.get('sql_host') or environment.get('sql_host')
        sql_adapter = options.get('sql_adapter') or 'mysql'


        cache_local_dir = ''
        if _option_true(options.get('enable-persistent-cache', 'true')):
            # Do not store this 'cache-local-dir' in the relstorage options.
            # We'll intermittently have buildout issues when writing this
            # to the installed.cfg while looking up the storage refs. We
            # avoid taking any user-defined values since it might be
            # confusing to have one (count limited) directory for all storages.
            cache_local_dir = '${deployment:cache-directory}/data_cache/${:name}.cache'

        # Utilizing the built in memcache capabilites is not
        # beneficial in all cases. In fact it rarely is. It's
        # semi-deprecated in RelStorage 3. If the recipe option
        # 'cache_servers' is empty or not defined, the relstorage
        # config options 'cache_module_name' and 'cache_module_name'
        # will be omitted from the generated config.
        cache_servers = options.get('cache_servers') or environment.get('cache_servers', '')
        if cache_servers.strip():
            extra_base_kwargs = {
                'cache_module_name': 'memcache',
                'cache_servers': cache_servers.strip()
            }
            remote_cache_config = ZConfigSnippet(**{
                k.replace('_', '-'): v
                for k, v
                in extra_base_kwargs.items()
            })
        else:
            extra_base_kwargs = {}
            remote_cache_config = ZConfigSnippet()

        relstorage_zcml = self.zlibstorage_wrapper(relstorage(remote_cache_config))
        filestorage_zcml = self.zlibstorage_wrapper(filestorage(self.ref('filestorage_name')))
        blob_cache_size = options.get('blob-cache-size', '')
        # Order matters
        base_storage_name = name + '_base_storage'

        base_storage_part = BaseStoragePart(
            base_storage_name,
            sql_user=sql_user,
            sql_passwd=sql_passwd,
            sql_host=sql_host,
            sql_adapter=sql_adapter,
            storage_zcml=relstorage_zcml,
            client_zcml=zodb(self.ref('name'), self.ref('storage_zcml')),
            filestorage_zcml=filestorage_zcml,
            relstorage_name_prefix=relstorage_name_prefix,
            cache_local_dir=cache_local_dir,
            blob_cache_size=blob_cache_size,
            **extra_base_kwargs
        )

        # TODO: Let this be configured for each storage.
        if not blob_cache_size:
            del base_storage_part['blob-cache-size']
            zcml = base_storage_part['storage_zcml']
            if hasattr(zcml, 'storage'):
                zcml = zcml.storage # unwrap zlibstorage
            del zcml['blob-cache-size']

        # TODO: This is for pool_timeout; it supports
        # configuring in _opts_base and _opts, but not per-storage.
        # Well, technically that's not true: per-storage is supported, but
        # there must be a a value in opts_base or _opts first.
        # We only document the shared values though.
        base_storage_part.buildout_lookup = self.make_buildout_lookup((
            name + '_opts_base',
            name + '_opts',
            extra_base_kwargs,
        ))

        self._parse(base_storage_part)
        storages = options['storages'].split()

        for storage in storages:
            part_name = name + '_' + storage.lower() + '_storage'
            # Note that while it would be nice to automatically extend
            # from this section, that leads to a recursive invocation
            # of this recipe, which obviously fails (with weird errors
            # about "part already exists"). So we use _opts for everything,
            # in precedence order
            other_bases_list = [
                base_storage_part,
                buildout.get(name + '_opts_base'),
                buildout.get(name + '_opts'),
                buildout.get(part_name + '_opts')
            ]
            part = Part(
                part_name,
                extends=other_bases_list,
                name=storage,
            )

            part = part.with_settings(**self.__adapter_settings(part))

            self._parse(part)

            self.create_directory(part_name, 'blob_dir')
            self.create_directory(part_name, 'cache-local-dir')
            self.add_database(part_name, 'client_zcml')

            if _option_true(options.get('write-zodbconvert', 'false')):
                self.__create_zodbconvert_parts(part)

        self.buildout_add_mkdirs(name='blob_dirs')
        self.buildout_add_zodb_conf()
        self.buildout_add_zeo_uris()

    def _resolve(self, part, obj):
        if isinstance(obj, SubstVar):
            if not obj.part: # Relative.
                return self._resolve(part, part.get(obj.setting))
            # buildout values are already fully resolved
            return self.buildout[obj.part][obj.setting] # pragma: no cover
        return obj


    def __clear_top_level_inherited_adapter_settings(self, sql_adapter_args):
        for k in BaseStoragePart.sql_adapter_args.keys():
            sql_adapter_args.pop(k, None)

    def _adapter_settings_for_sqlite3(self, part, sql_adapter_args):
        # sqlite resides on a single machine. No need to duplicate
        # blobs both in the DB and in the blob cache. This reduces parallel
        # commit, but it's not really parallel anyway.
        # Note that we DO NOT add the data-dir to the list of directories to create.
        # Uninstalling this part should not remove that directory, which is
        # what would happen if we added it.

        # Top-level settings which we got by default have to go; there are none.
        self.__clear_top_level_inherited_adapter_settings(sql_adapter_args)
        sql_adapter_args.addValue('data-dir', str(part['data_dir']) + '/' + part.name)
        return {
            'shared-blob-dir': True
        }

    def _adapter_settings_for_postgresql(self, part, sql_adapter_args):
        # If no DSN is specified in the sql_adapter_args then we compute one.
        if 'dsn' in sql_adapter_args:
            return {}
        # Hoist everything present by default into the dsn. Resolve them now so that we don't
        # put in empty fields and can use defaults.
        def resolve(obj):
            return self._resolve(part, obj)
        dsn = ' '
        for dsn_key, setting_key in (
                ('dbname', 'db'),
                ('user', 'user',),
                ('password', 'passwd'),
                ('host', 'host')
        ):
            setting = sql_adapter_args.pop(setting_key)[0]
            setting = resolve(setting)
            if setting:
                dsn += "%s='%s' " % (dsn_key, setting)
        dsn = dsn.strip()

        self.__clear_top_level_inherited_adapter_settings(sql_adapter_args)

        if 'sql_port' in sql_adapter_args:
            # Note no quotes
            dsn += ' port=%s' % (resolve(sql_adapter_args['sql_port'][0]))
            sql_adapter_args.pop('sql_port')

        sql_adapter_args.addValue('dsn', dsn)
        # No special settings to return, everything is in the mutated sql_adapter_args
        return {}

    def _adapter_settings_for_mysql(self, part, sql_adapter_args):
        # Our default is set up for MySQL
        return {}

    def __adapter_settings(self, part):
        # sql adapter args could be dict-like if its our default template,
        # or it could be a string if it's specified by the user to replace our default
        # template.
        sql_adapter_args = part['sql_adapter_args']
        if isinstance(sql_adapter_args, str):
            sql_adapter_args = ZConfig.schemaless.loadConfigFile(NativeStringIO(sql_adapter_args))
        else:
            sql_adapter_args = ZConfig.schemaless.Section(data={
                k: [v] for k, v in sql_adapter_args.items()
            })

        # Inline the sql_adapter_extra_args here so we can verify them as valid
        # and so the db-specific functions get an entire view
        extra_args = part['sql_adapter_extra_args']
        if hasattr(extra_args, 'const'):
            extra_args = extra_args.const
        if extra_args:
            config = ZConfig.schemaless.loadConfigFile(NativeStringIO(str(extra_args)))
            for k, v in config.items():
                sql_adapter_args[k] = v
            sql_adapter_args.sections.extend(config.sections)

        adapter_name = str(part.get('sql_adapter'))
        settings = getattr(self, '_adapter_settings_for_' + adapter_name)(part, sql_adapter_args)

        settings['sql_adapter_args'] = sql_adapter_args
        return settings

    def __create_zodbconvert_parts(self, part):
        # ZODB convert to and from files

        normalized_storage_name = part['name'].lower()

        src_part_name = 'zodbconvert_' + part.name + '_src'
        dest_part_name = 'zodbconvert_' + part.name + '_destination'
        self.create_directory(src_part_name, 'dump_dir')
        self.create_directory(dest_part_name, 'blob_dump_dir')

        to_relstorage_part_name = normalized_storage_name + '_to_relstorage_conf'
        from_relstorage_part_name = normalized_storage_name + '_from_relstorage_conf'

        src_part = Part(
            src_part_name,
            extends=part.extends,
            name='source',
            filestorage_name='destination',
            dump_name=normalized_storage_name,
            sql_db=part['name'],
        )
        src_part = src_part.with_settings(**self.__adapter_settings(part))
        self._parse(src_part)

        dest_part = src_part.named(dest_part_name).with_settings(
            name='destination',
            filestorage_name='source',
        )
        self._parse(dest_part)

        choices = {
            to_relstorage_part_name: dest_part_name,
            from_relstorage_part_name: src_part_name
        }
        to_relstorage_part = Part(
            to_relstorage_part_name,
            recipe='collective.recipe.template',
            output=Part.uses_name('${deployment:etc-directory}/relstorage/%s.xml'),
            input=[
                'inline:',
                self.zlibstorage_import(),
                '%import relstorage',
                self.choice_ref(choices, 'storage_zcml'),
                self.choice_ref(choices, 'filestorage_zcml'),
            ],
        )
        self._parse(to_relstorage_part)

        from_relstorage_part = to_relstorage_part.named(from_relstorage_part_name)
        self._parse(from_relstorage_part)
